### PR TITLE
Fix challenge computation

### PR DESCRIPTION
**Pull request by jules**
# What ❔

Fixes challenge computation to match the KZG spec.

## Checklist

- [x] PR title corresponds to the body of PR (we generate changelog entries from PRs).
- [x] Tests for the changes have been added / updated.
- [x] Documentation comments have been added / updated.
- [x] Code has been formatted via `zk fmt` and `zk lint`.

### CODE REVIEW

File: src/kzg/mod.rs [hunk 1]
```diff
@@ -305,14 +305,22 @@ fn eval_poly(blob: &[Fr], z: &Fr) -> Fr {
...
+    data.extend(&degree_separator);
+    let blob_bytes = blob
+        .iter()
+        .flat_map(|el| {
+            el.into_repr()
+                .0
+                .iter()
+                .rev()
+                .flat_map(|v| v.to_be_bytes())
+                .collect::<Vec<u8>>()
```
Comment (shamatar) [inline src/kzg/mod.rs]: you need a length assert here, and did they really decide to make a commitment based on the full blob bytes? This is strange move

File: src/kzg/mod.rs [hunk 2]
```diff
@@ -329,15 +337,12 @@ fn compute_challenge(blob: &[Fr], commitment: &G1Affine) -> Fr {
...
-    for (i, el) in ret.iter_mut().enumerate() {
-        *el |= bytes[i * 8] as u64;
-        *el |= (bytes[1 + i * 8] as u64) << 8;
-        *el |= (bytes[2 + i * 8] as u64) << 16;
-        *el |= (bytes[3 + i * 8] as u64) << 24;
-        *el |= (bytes[4 + i * 8] as u64) << 32;
-        *el |= (bytes[5 + i * 8] as u64) << 40;
-        *el |= (bytes[6 + i * 8] as u64) << 48;
-        *el |= (bytes[7 + i * 8] as u64) << 56;
+    for i in 0..ret.len() {
```
Comment (shamatar) [inline src/kzg/mod.rs]: we use nightly, so just do `array_chunks` instead

**Review by shamatar — CHANGES_REQUESTED** [review]

File: src/kzg/mod.rs (hunk 1 above)
Comment (jules) [reply src/kzg/mod.rs]: Yeah, as shown in the [consensus spec](https://github.com/ethereum/consensus-specs/blob/86fb82b221474cc89387fa6436806507b3849d88/specs/deneb/polynomial-commitments.md#compute_challenge) it requires the entire blob

File: src/kzg/mod.rs [hunk 3]
```diff
@@ -305,14 +305,23 @@ fn eval_poly(blob: &[Fr], z: &Fr) -> Fr {
...
-    let mut data = String::from("FSBLOBVERIFY_V1_");
-    let degree_separator: [u8; 16] = [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 10, 0];
-    data.push_str(std::str::from_utf8(&degree_separator).unwrap());
-    let encoded_blob = bincode::serialize(&blob).expect("should be able to serialize blob");
-    data.push_str(&format!("{:x?}", encoded_blob));
-    let encoded_commitment =
-        bincode::serialize(&commitment).expect("should be able to serialize commitment");
-    data.push_str(&format!("{:x?}", encoded_commitment));
+    let mut data = String::from("FSBLOBVERIFY_V1_").into_bytes();
+    let degree_separator: [u8; 16] = [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0x10, 0];
```
Comment (shamatar) [inline src/kzg/mod.rs:309]: Are you sure about endianess here?

File: src/kzg/mod.rs (hunk 1 above)
Comment (shamatar) [reply src/kzg/mod.rs]: Ok, but there is a much more efficient way to put all the bytes instead of tons on internal allocations

Comment (shamatar) [reply src/kzg/mod.rs]: And ideally you should just put everything into main buffer directly

File: src/kzg/mod.rs [hunk 4]
```diff
@@ -305,14 +305,23 @@ fn eval_poly(blob: &[Fr], z: &Fr) -> Fr {
...
+                .0
+                .iter()
+                .rev()
+                .flat_map(|v| v.to_be_bytes())
+                .collect::<Vec<u8>>()
+        })
+        .collect::<Vec<u8>>();
+    assert!(blob_bytes.len() == FIELD_ELEMENTS_PER_BLOB * 32);
+    data.extend(&blob_bytes);
+    data.extend(commitment.into_compressed().as_ref());
```
Comment (shamatar) [inline src/kzg/mod.rs:317]: Did you check that pairing's format of compression is the same as in 4844? It was quite custom

File: src/kzg/mod.rs [hunk 5]
```diff
@@ -329,15 +338,10 @@ fn compute_challenge(blob: &[Fr], commitment: &G1Affine) -> Fr {
 
 fn u8_repr_to_u64_repr(bytes: [u8; 32]) -> [u64; 4] {
```
Comment (shamatar) [inline src/kzg/mod.rs]: name endianess in function name

File: src/kzg/mod.rs [hunk 6]
```diff
@@ -329,15 +338,10 @@ fn compute_challenge(blob: &[Fr], commitment: &G1Affine) -> Fr {
...
-    for (i, el) in ret.iter_mut().enumerate() {
-        *el |= bytes[i * 8] as u64;
-        *el |= (bytes[1 + i * 8] as u64) << 8;
-        *el |= (bytes[2 + i * 8] as u64) << 16;
-        *el |= (bytes[3 + i * 8] as u64) << 24;
-        *el |= (bytes[4 + i * 8] as u64) << 32;
-        *el |= (bytes[5 + i * 8] as u64) << 40;
-        *el |= (bytes[6 + i * 8] as u64) << 48;
-        *el |= (bytes[7 + i * 8] as u64) << 56;
+    for (i, chunk) in bytes.array_chunks::<8>().enumerate() {
```
Comment (shamatar) [inline src/kzg/mod.rs]: Please less index use, more "zip" and "iter_mut()" for destination

File: src/kzg/mod.rs (hunk 3 above)
Comment (jules) [reply src/kzg/mod.rs:309]: yes, it was compared to onchain data, the degree separator is big endian

File: src/kzg/mod.rs (hunk 4 above)
Comment (jules) [reply src/kzg/mod.rs:317]: yes, also checked against onchain data - the default function here passed and no need to swap endianness

**Review by koloz193 — APPROVED** [review]
tested against mainnet blob/commitment and tests are passing

**Pull request merged**